### PR TITLE
Implement window.openURLInDefaultBrowser() (fixes #11292)

### LINKED ISSUE #11292: Implement window.openURLInDefaultBrowser()
We want to be able to open links in the default browser. This will be used for the crash reporter in browserhtml (see https://github.com/servo/servo/pull/11286 and https://github.com/browserhtml/browserhtml/issues/1001 for context).

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11465)

### CODE REVIEW

File: components/script/dom/webidls/BrowserElement.webidl [hunk 1]
```diff
@@ -146,24 +146,16 @@ interface BrowserElementPrivileged {
...
-  [Throws,
-   Pref="dom.mozbrowser.enabled",
-   CheckAnyPermissions="browser"]
+  [Func="Window::global_is_mozbrowser", Throws]
   void reload(optional boolean hardReload = false);
 
-  [Throws,
-   Pref="dom.mozbrowser.enabled",
-   CheckAnyPermissions="browser"]
+  [Func="Window::global_is_mozbrowser", Throws]
```
Comment (jdm) [inline components/script/dom/webidls/BrowserElement.webidl]: We cold probably add a single Func to the interface, instead.

File: components/script/dom/webidls/Window.webidl [hunk 2]
```diff
@@ -159,6 +159,8 @@ partial interface Window {
   void debug(DOMString arg);
   void gc();
   void trap();
+  [Func="Window::global_is_mozbrowser", Throws]
+  void openURLInDefaultBrowser(DOMString href);
```
Comment (jdm) [inline components/script/dom/webidls/Window.webidl]: Why not add this to BrowserPrivileged?

File: components/script/dom/window.rs [hunk 3]
```diff
@@ -838,6 +839,15 @@ impl WindowMethods for Window {
     // https://html.spec.whatwg.org/multipage/#dom-window-status
     fn SetStatus(&self, status: DOMString) {
         *self.status.borrow_mut() = status
+
+    fn OpenURLInDefaultBrowser(&self, href: DOMString) -> ErrorResult {
```
Comment (jdm) [inline components/script/dom/window.rs]: Tidy's going to complain about a spec link.

File: components/script/lib.rs [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #![feature(optin_builtin_traits)]
 #![feature(peekable_is_empty)]
 #![feature(plugin)]
+#![feature(question_mark)]
```
Comment (jdm) [inline components/script/lib.rs]: Where does this get used?

**Comment by nox** [comment]
-S-needs-code-changes +S-awaiting-answer 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/script/lib.rs, line 20 \[r6\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9b--r6-20:-KImFc5TXb3s_BKU7_Hc:1753122725) ([raw file](https://github.com/servo/servo/blob/3b9d25391b7afdb1aec335ce0c9d55efeec3ed3b/components/script/lib.rs#L20)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Where does this get used?
</details>
In `Window::OpenURLInDefaultBrowser`.

---

*[components/script/dom/window.rs, line 843 \[r6\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9az-r6-843:-KImFewkG6UGblC_kbGI:747064767) ([raw file](https://github.com/servo/servo/blob/3b9d25391b7afdb1aec335ce0c9d55efeec3ed3b/components/script/dom/window.rs#L843)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Tidy's going to complain about a spec link.
</details>
What do I do?

---

*[components/script/dom/webidls/BrowserElement.webidl, line 158 \[r5\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9aw-r5-158:-KImFg8nv1QWb5wjZFku:-1258773187) ([raw file](https://github.com/servo/servo/blob/f285e2bd2b5608922e8fdeef884fe38c79eeb5bd/components/script/dom/webidls/BrowserElement.webidl#L158)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> We cold probably add a single Func to the interface, instead.
</details>
Doesn't work.

```
Interface with no interface object is exposed conditionally
```

---

*[components/script/dom/webidls/Window.webidl, line 163 \[r6\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9ay-r6-163:-KImFjFKaUjJegtRN5FP:952744085) ([raw file](https://github.com/servo/servo/blob/3b9d25391b7afdb1aec335ce0c9d55efeec3ed3b/components/script/dom/webidls/Window.webidl#L163)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Why not add this to BrowserPrivileged?
</details>
Because `Window` does not implement `BrowserElementPrivileged`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11465#-:-KImFpTdRWB8PM7jChtk:-1162193221)*

**Comment by jdm** [comment]
-S-awaiting-answer +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 8 of 8 files at r4, 1 of 1 files at r5, 7 of 7 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/lib.rs, line 20 \[r6\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9b--r6-20:-KImWeKhFsvv7FY25Doq:1420878687) ([raw file](https://github.com/servo/servo/blob/3b9d25391b7afdb1aec335ce0c9d55efeec3ed3b/components/script/lib.rs#L20)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> In `Window::OpenURLInDefaultBrowser`.
</details>
Let's take it out for the purpose of this PR.

---

*[components/script/dom/window.rs, line 843 \[r6\]](https://reviewable.io:443/reviews/servo/servo/11465#-KIm7Cgu3W7noL6Nw9az-r6-843:-KImUuuVzq4Lt5HdueBl:-1893203141) ([raw file](https://github.com/servo/servo/blob/3b9d25391b7afdb1aec335ce0c9d55efeec3ed3b/components/script/dom/window.rs#L843)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> What do I do?
</details>
// check-tidy: no specs after this line

---

*[components/script/dom/window.rs, line 847 \[r7\]](https://reviewable.io:443/reviews/servo/servo/11465#-KImX1tyNFImyL3jYrjw:-KImX1tzzyJuCWLWIs8h:-1230773089) ([raw file](https://github.com/servo/servo/blob/10d1451f344027c284612263737f7a64ce27874b/components/script/dom/window.rs#L847)):*
> ```Rust
>         let url = Url::parse(&href).map_err(|e| {
>             Error::Type(format!("Couldn't parse URL: {}", e))
>         })?;
> ```

I'd like to have a discussion about whether we want to use ? before we start introducing it, so let's use try! here instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11465#-:-KImX5wwnzOrac1OGzoF:-1847720821)*

**Comment by jdm** [comment]
@bors-servo: r+

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 2 of 2 files at r8, 8 of 8 files at r9, 8 of 8 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11465#-:-KImcKvES9ucLK0D_SJT:-317921516)*

**Pull request merged**